chirp/drivers/baofeng_uv17Pro.py

```diff
@@ -0,0 +1,1427 @@
+            for j in range(0, length):
+                try:
+                    code.append(DTMF_CHARS.index(str(setting.value)[j]))
+                except IndexError:
+                    code.append(0xFF)
+            obj.code = code
+
+        for i in range(0, 20):
+            _codeobj = self._memobj.pttid[i].code
+            _code = "".join([
+                DTMF_CHARS[x] for x in _codeobj if int(x) < 0x1F])
+            val = RadioSettingValueString(0, 5, _code, False)
+            val.set_charset(DTMF_CHARS)
+            pttid = RadioSetting("pttid/%i.code" % i,
+                                 "Signal Code %i" % (i + 1), val)
+            pttid.set_apply_callback(apply_code, self._memobj.pttid[i], 5)
+            dtmfe.append(pttid)
+
+        if _mem.ani.dtmfon > 0xC3:
+            val = 0x03
+        else:
+            val = _mem.ani.dtmfon
+        rs = RadioSetting("ani.dtmfon", "DTMF Speed (on)",
+                          RadioSettingValueList(LIST_DTMFSPEED,
+                                                LIST_DTMFSPEED[val]))
+        dtmfe.append(rs)
+
+        if _mem.ani.dtmfoff > 0xC3:
+            val = 0x03
+        else:
+            val = _mem.ani.dtmfoff
+        rs = RadioSetting("ani.dtmfoff", "DTMF Speed (off)",
+                          RadioSettingValueList(LIST_DTMFSPEED,
+                                                LIST_DTMFSPEED[val]))
+        dtmfe.append(rs)
+
+        _codeobj = self._memobj.ani.code
+        _code = "".join([DTMF_CHARS[x] for x in _codeobj if int(x) < 0x1F])
+        val = RadioSettingValueString(0, 5, _code, False)
+        val.set_charset(DTMF_CHARS)
+        rs = RadioSetting("ani.code", "ANI Code", val)
+        rs.set_apply_callback(apply_code, self._memobj.ani, 5)
+        dtmfe.append(rs)
+
+        if self._aniid:
+            rs = RadioSetting("ani.aniid", "When to send ANI ID",
+                              RadioSettingValueList(LIST_PTTID,
+                                                    LIST_PTTID[
+                                                        _mem.ani.aniid]))
+            dtmfe.append(rs)
+
+        rs = RadioSetting("settings.hangup", "Hang-up time",
+                          RadioSettingValueList(LIST_HANGUPTIME,
+                                                LIST_HANGUPTIME[
+                                                    _mem.settings.hangup]))
+        dtmfe.append(rs)
+
+        def convert_bytes_to_freq(bytes):
+            real_freq = 0
+            for byte in bytes:
+                real_freq = (real_freq * 10) + byte
+            return chirp_common.format_freq(real_freq * 10)
+
+        def my_validate(value):
+            value = chirp_common.parse_freq(value)
+            freqOk = False
+            for band in self.VALID_BANDS:
+                if value > band[0] and value < band[1]:
+                    freqOk = True
+            if not freqOk:
+                raise InvalidValueError("Invalid frequency!")
```

Comment:
Doing this will make it so that the user will not be able to see any settings at all and will just get this error when they click on that tab. It won't tell them which frequency is invalid and there will be no way for them to fix it. You should probably just display it anyway, log something, etc. Perhaps just change it to the lower bound of VALID_BANDS?

Reply:
The way it is coded now, it just reverts to the frequency that was already in the vfo field. It does not affect the settings tab. If the error is raised, the field reverts to its previous value. It's done this way in many baofeng drivers.
I did make the error more verbose now.

Reply:
You're assuming you only hit that error if the user enters an incorrect frequency. However, if it's already bad in the radio, then you'll break everything. We see a ton of examples of the chinese radios coming from the factory with seemingly corrupt defaults and as you can see from the raft of bugs we have, it's very annoying for people to get no settings and be unable to fix them. So please just coerce the value here. Even if they put in something invalid and have it snap back to the nearest value, that should be okay. I again said these VFO things should probably be special channels so the UI will do all the normal valid_bands enforcement for you, like this:

<img width="427" alt="image" src="https://github.com/kk7ds/chirp/assets/3046399/f312aacd-83a8-459a-8b98-6bd70ea7d309">

Personally I see zero reason to go to the work of exposing the VFOs in the programming software. I've never once felt like it would be helpful to do that in my own programming. So if it were me, I'd nix all that stuff from here.

Reply:
Fixed: removed VFO settings.